### ISSUE TITLE
DELG implementation does not have the auto-encoder as cited in paper.

### ISSUE THREAD

**Issue by wzmJimmy** [tensorflow/models#9189]
## The DELG paper

According to the latest version [DELG paper](https://arxiv.org/pdf/2001.05027.pdf), the local feature part of DELG is slightly different from DELF. It apply an auto-encoder on features and then use the attention weights on reconstructed features. losses of  auto-encoder reconstruction and attention classifier are used together to train the local features. Just like the two screenshot below:

<img width="507" alt="截屏2020-09-02 下午7 23 51" src="https://user-images.githubusercontent.com/35741367/91975641-4c371180-ed52-11ea-80de-a1e239c00ea3.png">

<img width="552" alt="截屏2020-09-02 下午7 21 35" src="https://user-images.githubusercontent.com/35741367/91975675-548f4c80-ed52-11ea-875f-afda26274b52.png">

## The implementation of DELG in DELF repo

**However, the code implementation of DELG here does not include the auto-encoder.**

1. In [delg_model.py](https://github.com/tensorflow/models/blob/master/research/delf/delf/python/training/model/delg_model.py), The class  DELG is a subclass of DELF and use the exact local feature part as DELF.

2. in [train.py](https://github.com/tensorflow/models/blob/master/research/delf/delf/python/training/train.py), we can look at the train_step function (line 242 -322).  Only SparseCategoricalCrossentropy losses of two classifier (descriptor and attention) are taken into account.

## Question about the auto-encoder feature

Does the implementation of DELG here miss the auto-encoder? or the paper has a wrong description about its model structure?

Is there any plan to add the auto-encoder inside?

**Comment by andrefaraujo** [comment]
Yes, currently the auto-encoder is missing, you are right. We are planning to add it soonish, hopefully within the coming month.

**Comment by wzmJimmy** [comment]
That's great, hope it will come out soon.


Moreover, I have another question. **The paper itself has conflicted descriptions about which is the local descriptor**:

<img width="640" alt="截屏2020-09-03 上午4 12 21" src="https://user-images.githubusercontent.com/35741367/92031984-0868fa00-ed9c-11ea-8131-b598afcde98b.png">

- According to the graph above (The 3.2 Model section), the local descriptor should be the encoded low-dimension representation L. 

- While the picture in my original post (the 3.3 Training section) says they  use the reconstructed feature S' as local descriptor.

In my opinion, the first one makes more sense. It's kind of waste if we use reconstructed feature S'  of AE as local descriptor and train AE with a MSE loss of reconstruction, Why not directly use the original S?


**Which one is the one used by the paper? What's the exact setting?**  

My understanding is that AE is applied after the subsampled Block3 output, which is the same input as the attention model. The encoder is a 1x1 conv2D layer with 128 output filters, while the decoder is another 1x1 conv2D layer with 1024 output and Relu activation. The encoded L is local descriptor and reconstructed S' is for MSE loss computing.  Is that correct? 

if this is the case, is there any activation over encoder?

**Comment by andrefaraujo** [comment]
The local descriptor is the encoded low-dimension representation. I didn't see it mentioning "the reconstructed feature S' as local descriptor" as you mentioned.

> My understanding is that AE is applied after the subsampled Block3 output, which is the same input as the attention model. The encoder is a 1x1 conv2D layer with 128 output filters, while the decoder is another 1x1 conv2D layer with 1024 output and Relu activation. The encoded L is local descriptor and reconstructed S' is for MSE loss computing. Is that correct?

Correct!

> if this is the case, is there any activation over encoder?

No, there is no activation function there. The paper hints at this: "Note that, contrary to S, the local descriptors L are not restricted to be non-negative."

**Comment by wzmJimmy** [comment]
Thank you for helping me explain it.

I just found that I misunderstood the first two pictures. They use encoded low-dimension representation as local descriptor as well. Whereas, if that's the case. Why does it pool on the reconstructed feature S' with attention weight before attention classifier, rather than pool directly on the local descriptor? That could help the local descriptor capture more relevant information.

Moreover, why does the attention classifier not use an Arc-face like the global classifier, but use the standard softmax-cross-entropy? Is there any benefit of doing this way?

**Comment by andrefaraujo** [comment]
1) We found that pooling on the local descriptor directly can be difficult, because it will lead to a lower dimensionality pooled feature, which may have a difficult time optimizing the classifier (consequently, the attention layers cannot be learned so well). Also, if we allow the local descriptors to be tuned directly based on this loss, it may produce abstract/high-level representations which may be good for the attention loss optimization, but not necessarily for local descriptor matching (the local descriptors may become less localizable).

2) Attention classifier should be able to reuse ArcFace as well; it requires an additional hyperparameter to be set though (the ArcFace margin). The goal of the attention layer is not to produce a powerful global feature, but rather to learn well the attention keypoint detection; so the ArcFace loss may not contribute much to this goal.

**Comment by sky186** [comment]
@andrefaraujo 
1、Hello , I have a question ,the attention layer , two conv ,out filter is 1 , not 512 ?   
"self.conv2 = layers.Conv2D(
        1,"
`
class AttentionModel(tf.keras.Model):
  """Instantiates attention model.
  Uses two [kernel_size x kernel_size] convolutions and softplus as activation
  to compute an attention map with the same resolution as the featuremap.
  Features l2-normalized and aggregated using attention probabilites as weights.
  """

  def __init__(self, kernel_size=1, decay=_DECAY, name='attention'):
    """Initialization of attention model.
    Args:
      kernel_size: int, kernel size of convolutions.
      decay: float, decay for l2 regularization of kernel weights.
      name: str, name to identify model.
    """
    super(AttentionModel, self).__init__(name=name)

    # First convolutional layer (called with relu activation).
    self.conv1 = layers.Conv2D(
        512,
        kernel_size,
        kernel_regularizer=reg.l2(decay),
        padding='same',
        name='attn_conv1')
    self.bn_conv1 = layers.BatchNormalization(axis=3, name='bn_conv1')

    # Second convolutional layer, with softplus activation.
    self.conv2 = layers.Conv2D(
        1,
        kernel_size,
        kernel_regularizer=reg.l2(decay),
        padding='same',
        name='attn_conv2')
    self.activation_layer = layers.Activation('softplus')
`

2、now，not autoencoder layer  and restruction loss ?

**Comment by andrefaraujo** [comment]
1) 512 is the output of conv1, as you can see. Final output is 1D score per feature, which is the keypoint detection score.
2) Correct, the autoencoder and reconstruction loss are not yet implemented.

**Comment by HouBiaoLiu** [comment]
If we use DELG in vehicleID datasets that contain frontal and back directions,  Dose it have effect?

**Comment by brejchajan** [comment]
Hi, thank you very much for the DELG implementation. I would like to kindly ask what the status of the DELG auto-encoder is?

**Comment by andrefaraujo** [comment]
Done, we just checked it in. Thanks for the patience: https://github.com/tensorflow/models/pull/9555

**Comment by kamwoh** [comment]
Hi, although the autoencoder is implemented, the provided pre-trained weights of DELG R50 GLDv2 contains a batch normalization layer for autoencoder but the implementation doesn't.

The snippet below is part of the weights from `r50delg_gldv2clean_20200914` 
```
<tf.Variable 'dim_reduction/conv1/weights:0' shape=(1, 1, 1024, 128) dtype=float32_ref>
<tf.Variable 'dim_reduction/conv1/BatchNorm/gamma:0' shape=(128,) dtype=float32_ref>
<tf.Variable 'dim_reduction/conv1/BatchNorm/beta:0' shape=(128,) dtype=float32_ref>
<tf.Variable 'dim_reduction/conv1/BatchNorm/moving_mean:0' shape=(128,) dtype=float32_ref>
<tf.Variable 'dim_reduction/conv1/BatchNorm/moving_variance:0' shape=(128,) dtype=float32_ref>
<tf.Variable 'dim_expand/conv1/weights:0' shape=(1, 1, 128, 1024) dtype=float32_ref>
<tf.Variable 'dim_expand/conv1/BatchNorm/gamma:0' shape=(1024,) dtype=float32_ref>
<tf.Variable 'dim_expand/conv1/BatchNorm/beta:0' shape=(1024,) dtype=float32_ref>
<tf.Variable 'dim_expand/conv1/BatchNorm/moving_mean:0' shape=(1024,) dtype=float32_ref>
<tf.Variable 'dim_expand/conv1/BatchNorm/moving_variance:0' shape=(1024,) dtype=float32_ref>
```

**Comment by andrefaraujo** [comment]
Great catch, @kamwoh -- thanks!

What we need is to add a BatchNormalization layer (eg, [like done in the DELF Attention part](https://github.com/tensorflow/models/blob/master/research/delf/delf/python/training/model/delf_model.py#L59)) to the [AutoEncoder](https://github.com/tensorflow/models/blob/master/research/delf/delf/python/training/model/delf_model.py#L89). Seems quite simple overall. Would you like to send me a Pull Request, @kamwoh ? I'd be happy to review and integrate it.
